src/actors/util/actor_util.go

```diff
@@ -1,10 +1,95 @@
 package util
 
 import (
+	"math/big"
+
 	addr "github.com/filecoin-project/go-address"
 	abi "github.com/filecoin-project/specs/actors/abi"
 )
 
+type BalanceTableHAMT map[addr.Address]abi.TokenAmount
+
+type DealIDSetHAMT map[abi.DealID]bool
+type IntToDealIDHAMT map[int64]abi.DealID
+
+type DealIDQueue struct {
+	Values     IntToDealIDHAMT
+	StartIndex int64
+	EndIndex   int64
+}
+
+func (x *DealIDQueue) Enqueue(dealID abi.DealID) {
+	nextIndex := x.EndIndex
+	x.Values[nextIndex] = dealID
+	x.EndIndex = nextIndex + 1
+}
+
+func (x *DealIDQueue) Dequeue() (dealID abi.DealID, ok bool) {
+	AssertMsg(x.StartIndex <= x.EndIndex, "index %d > end %d", x.StartIndex, x.EndIndex)
+
+	if x.StartIndex == x.EndIndex {
+		dealID = abi.DealID(-1)
+		ok = false
+		return
+	} else {
+		dealID = x.Values[x.StartIndex]
+		delete(x.Values, x.StartIndex)
+		x.StartIndex += 1
+		ok = true
+		return
+	}
+}
+
+func DealIDQueue_Empty() DealIDQueue {
+	return DealIDQueue{
+		Values:     IntToDealIDHAMT_Empty(),
+		StartIndex: 0,
+		EndIndex:   0,
+	}
+}
+
+type MinerSetHAMT map[addr.Address]bool
+
+type ActorIDSetHAMT map[abi.ActorID]bool
+
+type MinerEventInterface interface {
+	MinerAddr() addr.Address
+	Sectors() []abi.SectorNumber
+}
+
+type MinerEvent struct {
+	MinerAddress addr.Address
+	MinerSectors []abi.SectorNumber // Empty for global events, such as SurprisePoSt expiration.
+}
+
+func (m MinerEvent) MinerAddr() addr.Address {
+	return m.MinerAddress
+}
+
+func (m MinerEvent) Sectors() []abi.SectorNumber {
+	return m.MinerSectors
+}
+
+// TODO using an interface in a map is probably a bad idea since different implementations may provide incomparable
+// type, but for now lets do the easy thing
+type MinerEventSetHAMT map[MinerEventInterface]bool
```

Comment:
Will this suffice until we complete the extraction?

Reply:
No, we can't have any interfaces in these on-chain data types. Remove the interface entirely.

Reply:
For now put this in a slice and deal with the slowness.